src/mon/MonClient.h

```diff
@@ -380,55 +425,181 @@ class MonClient : public Dispatcher {
 
   struct MonCommand {
     string target_name;
-    int target_rank;
+    int target_rank = -1;
     uint64_t tid;
     vector<string> cmd;
     bufferlist inbl;
-    bufferlist *poutbl;
-    string *prs;
-    int *prval;
-    Context *onfinish, *ontimeout;
-
-    explicit MonCommand(uint64_t t)
-      : target_rank(-1),
-	tid(t),
-	poutbl(NULL), prs(NULL), prval(NULL), onfinish(NULL), ontimeout(NULL)
-    {}
+    std::unique_ptr<CommandCompletion> onfinish;
+    std::optional<boost::asio::steady_timer> cancel_timer;
+
+    MonCommand(MonClient& monc, uint64_t t,
+	       std::unique_ptr<CommandCompletion> onfinish)
+      : tid(t), onfinish(std::move(onfinish)) {
+      auto timeout = ceph::maybe_timespan(monc.cct->_conf->rados_mon_op_timeout);
+      if (timeout) {
+	cancel_timer.emplace(monc.service, *timeout);
+	cancel_timer->async_wait(
+          [this, &monc](boost::system::error_code ec) {
+	    if (ec)
+	      return;
+	    Mutex::Locker l(monc.monc_lock);
+	    monc._cancel_mon_command(tid);
+	  });
+      }
+    }
   };
+  friend MonCommand;
   map<uint64_t,MonCommand*> mon_commands;
 
   void _send_command(MonCommand *r);
   void _resend_mon_commands();
   int _cancel_mon_command(uint64_t tid);
-  void _finish_command(MonCommand *r, int ret, string rs);
+  void _finish_command(MonCommand *r, boost::system::error_code ret, string&& rs, bufferlist&& bl);
```

Comment:
This is ok, but string_view and a copy seems like it should also be ok-- since later on it's being called with const char* and constructing a std::string anyway, from what I can see.